// @mysten/ledgerjs-hw-app-sui/dist/Sui.mjs
import { buildDescriptor } from "./descriptor.mjs";
import { DeviceModelId } from "@ledgerhq/devices";
import { LatestFirmwareVersionRequired, UpdateYourApp } from "@ledgerhq/errors";
import { loadPKI } from "@ledgerhq/hw-bolos";
import { TransportStatusError } from "@ledgerhq/hw-transport";
import calService from "@ledgerhq/ledger-cal-service";
import sha256 from "fast-sha256";
import semver from "semver";

//#region src/Sui.ts
var LedgerToHost = /* @__PURE__ */ function(LedgerToHost$1) {
	LedgerToHost$1[LedgerToHost$1["RESULT_ACCUMULATING"] = 0] = "RESULT_ACCUMULATING";
	LedgerToHost$1[LedgerToHost$1["RESULT_FINAL"] = 1] = "RESULT_FINAL";
	LedgerToHost$1[LedgerToHost$1["GET_CHUNK"] = 2] = "GET_CHUNK";
	LedgerToHost$1[LedgerToHost$1["PUT_CHUNK"] = 3] = "PUT_CHUNK";
	return LedgerToHost$1;
}(LedgerToHost || {});
var HostToLedger = /* @__PURE__ */ function(HostToLedger$1) {
	HostToLedger$1[HostToLedger$1["START"] = 0] = "START";
	HostToLedger$1[HostToLedger$1["GET_CHUNK_RESPONSE_SUCCESS"] = 1] = "GET_CHUNK_RESPONSE_SUCCESS";
	HostToLedger$1[HostToLedger$1["GET_CHUNK_RESPONSE_FAILURE"] = 2] = "GET_CHUNK_RESPONSE_FAILURE";
	HostToLedger$1[HostToLedger$1["PUT_CHUNK_RESPONSE"] = 3] = "PUT_CHUNK_RESPONSE";
	HostToLedger$1[HostToLedger$1["RESULT_ACCUMULATING_RESPONSE"] = 4] = "RESULT_ACCUMULATING_RESPONSE";
	return HostToLedger$1;
}(HostToLedger || {});
const MIN_VERSION = "1.2.2";
const MANAGER_APP_NAME = "Sui";
function isPKIUnsupportedError(err) {
	return err instanceof TransportStatusError && err.message.includes("0x6a81");
}
async function tryLoadPKI(...args) {
	try {
		await loadPKI(...args);
	} catch (err) {
		if (isPKIUnsupportedError(err)) throw new LatestFirmwareVersionRequired("LatestFirmwareVersionRequired");
	}
}
/**
* Sui API
*
* @example
* import Sui from "@mysten/ledgerjs-hw-app-sui";
* const sui = new Sui(transport)
*/
var Sui = class {
	#verbose;
	constructor(transport, scrambleKey = "default_sui_scramble_key", verbose = false) {
		this.#verbose = verbose;
		this.transport = transport;
		this.transport.decorateAppAPIMethods(this, [
			"getPublicKey",
			"signTransaction",
			"getVersion",
			"provideTrustedDynamicDescriptor"
		], scrambleKey);
	}
	/**
	* Retrieves the public key associated with a particular BIP32 path from the Ledger app.
	*
	* @param path - the path to retrieve.
	* @param displayOnDevice - whether or not the address should be displayed on the device.
	*
	*/
	async getPublicKey(path, displayOnDevice = false) {
		const cla = 0;
		const ins = displayOnDevice ? 1 : 2;
		const p1 = 0;
		const p2 = 0;
		const payload = buildBip32KeyPayload(path);
		const response = await this.#sendChunks(cla, ins, p1, p2, payload);
		const keySize = response[0];
		const publicKey = response.slice(1, keySize + 1);
		let address = null;
		if (response.length > keySize + 2) {
			const addressSize = response[keySize + 1];
			address = response.slice(keySize + 2, keySize + 2 + addressSize);
		}
		return {
			publicKey,
			address
		};
	}
	/**
	* Sign a transaction with the key at a BIP32 path.
	*
	* @param txn - The transaction bytes to sign.
	* @param path - The path to use when signing the transaction.
	* @param options - Additional options used for clear signing purposes.
	* @param resolution - Additional data for token clear signing purposes.
	*/
	async signTransaction(path, txn, options, resolution) {
		const cla = 0;
		const ins = 3;
		const p1 = 0;
		const p2 = 0;
		if (this.#verbose) this.#log(txn);
		const rawTxn = Buffer.from(txn);
		const hashSize = Buffer.alloc(4);
		hashSize.writeUInt32LE(rawTxn.length, 0);
		const payloadTxn = Buffer.concat([hashSize, rawTxn]);
		this.#log("Payload Txn", payloadTxn);
		const payloads = [payloadTxn, buildBip32KeyPayload(path)];
		const { major } = await this.#internalGetVersion();
		this.#log("App version", major);
		const bcsObjects = options?.bcsObjects ?? [];
		this.#log("Objects list length", bcsObjects.length);
		if (major > 0 && bcsObjects.length > 0) {
			const numItems = Buffer.alloc(4);
			numItems.writeUInt32LE(bcsObjects.length, 0);
			let listData = Buffer.from(numItems);
			for (const item of bcsObjects) {
				const rawItem = Buffer.from(item);
				const itemLen = Buffer.alloc(4);
				itemLen.writeUInt32LE(rawItem.length, 0);
				listData = Buffer.concat([
					listData,
					itemLen,
					rawItem
				]);
			}
			payloads.push(listData);
		}
		if (resolution) {
			if (!resolution.deviceModelId) throw new Error("Resolution provided without a deviceModelId");
			if (resolution.deviceModelId !== DeviceModelId.nanoS) {
				await this.#checkAppVersion(MIN_VERSION, { throwOnOutdated: true });
				const { descriptor, signature } = await calService.getCertificate(resolution.deviceModelId, "trusted_name", "latest", { signatureKind: resolution.certificateSignatureKind });
				try {
					await loadPKI(this.transport, "TRUSTED_NAME", descriptor, signature);
				} catch (err) {
					if (isPKIUnsupportedError(err)) throw new LatestFirmwareVersionRequired("LatestFirmwareVersionRequired");
				}
				if (resolution.tokenId) {
					const { descriptor: coinMetaDescriptor, signature: coinMetaSignature } = await calService.getCertificate(resolution.deviceModelId, "coin_meta", "latest", { signatureKind: resolution.certificateSignatureKind });
					await tryLoadPKI(this.transport, "COIN_META", coinMetaDescriptor, coinMetaSignature);
					const token = await calService.findToken({ id: resolution.tokenId }, { signatureKind: resolution.certificateSignatureKind });
					await this.#provideTrustedDynamicDescriptor({
						data: Buffer.from(token.descriptor.data, "hex"),
						signature: Buffer.from(token.descriptor.signature, "hex")
					});
				}
			}
		}
		return { signature: await this.#sendChunks(cla, ins, p1, p2, payloads) };
	}
	/**
	* Retrieve the app version on the attached Ledger device.
	*/
	async #checkAppVersion(minVersion, { throwOnOutdated }) {
		const { major, minor, patch } = await this.#internalGetVersion();
		if (semver.lt(`${major}.${minor}.${patch}`, minVersion) && throwOnOutdated) throw new UpdateYourApp(void 0, { managerAppName: MANAGER_APP_NAME });
	}
	/**
	* Retrieve the app version on the attached Ledger device.
	*/
	async getVersion() {
		return await this.#internalGetVersion();
	}
	async #internalGetVersion() {
		const [major, minor, patch] = await this.#sendChunks(0, 0, 0, 0, Buffer.alloc(1));
		return {
			major,
			minor,
			patch
		};
	}
	/**
	* Provides trusted dynamic and signed coin metadata
	*
	* @param data An object containing the descriptor and its signature from the CAL
	*/
	async #provideTrustedDynamicDescriptor(data) {
		await this.#sendChunks(0, 34, 0, 0, buildDescriptor(data));
		return true;
	}
	/**
	* Convert a raw payload into what is essentially a singly-linked list of chunks, which
	* allows the ledger to re-seek the data in a secure fashion.
	*/
	async #sendChunks(cla, ins, p1, p2, payload, extraData = /* @__PURE__ */ new Map()) {
		const chunkSize = 180;
		if (!(payload instanceof Array)) payload = [payload];
		const parameterList = [];
		let data = new Map(extraData);
		for (let j = 0; j < payload.length; j++) {
			const chunkList = [];
			for (let i = 0; i < payload[j].length; i += chunkSize) {
				const cur = payload[j].slice(i, i + chunkSize);
				chunkList.push(cur);
			}
			let lastHash = Buffer.alloc(32);
			this.#log(lastHash);
			data = chunkList.reduceRight((blocks, chunk) => {
				const linkedChunk = Buffer.concat([lastHash, chunk]);
				this.#log("Chunk: ", chunk);
				this.#log("linkedChunk: ", linkedChunk);
				lastHash = Buffer.from(sha256(linkedChunk));
				blocks.set(lastHash.toString("hex"), linkedChunk);
				return blocks;
			}, data);
			parameterList.push(lastHash);
			lastHash = Buffer.alloc(32);
		}
		this.#log(data);
		return await this.#handleBlocksProtocol(cla, ins, p1, p2, Buffer.concat([Buffer.from([HostToLedger.START])].concat(parameterList)), data);
	}
	async #handleBlocksProtocol(cla, ins, p1, p2, initialPayload, data) {
		let payload = initialPayload;
		let result = Buffer.alloc(0);
		do {
			this.#log("Sending payload to ledger: ", payload.toString("hex"));
			const rv = await this.transport.send(cla, ins, p1, p2, payload);
			this.#log("Received response: ", rv);
			var rv_instruction = rv[0];
			const rv_payload = rv.slice(1, rv.length - 2);
			if (!(rv_instruction in LedgerToHost)) throw new TypeError("Unknown instruction returned from ledger");
			switch (rv_instruction) {
				case LedgerToHost.RESULT_ACCUMULATING:
				case LedgerToHost.RESULT_FINAL:
					result = Buffer.concat([result, rv_payload]);
					payload = Buffer.from([HostToLedger.RESULT_ACCUMULATING_RESPONSE]);
					break;
				case LedgerToHost.GET_CHUNK:
					const chunk = data.get(rv_payload.toString("hex"));
					this.#log("Getting block ", rv_payload);
					this.#log("Found block ", chunk);
					if (chunk) payload = Buffer.concat([Buffer.from([HostToLedger.GET_CHUNK_RESPONSE_SUCCESS]), chunk]);
					else payload = Buffer.from([HostToLedger.GET_CHUNK_RESPONSE_FAILURE]);
					break;
				case LedgerToHost.PUT_CHUNK:
					data.set(Buffer.from(sha256(rv_payload)).toString("hex"), rv_payload);
					payload = Buffer.from([HostToLedger.PUT_CHUNK_RESPONSE]);
					break;
			}
		} while (rv_instruction !== LedgerToHost.RESULT_FINAL);
		return result;
	}
	#log(...args) {
		if (this.#verbose) console.log(args);
	}
};
function buildBip32KeyPayload(path) {
	const paths = splitPath(path);
	const payload = Buffer.alloc(1 + paths.length * 4);
	payload[0] = paths.length;
	paths.forEach((element, index) => {
		payload.writeUInt32LE(element, 1 + 4 * index);
	});
	return payload;
}
function splitPath(path) {
	const result = [];
	path.split("/").forEach((element) => {
		let number = parseInt(element, 10);
		if (isNaN(number)) return;
		if (element.length > 1 && element[element.length - 1] === "'") number += 2147483648;
		result.push(number);
	});
	return result;
}

//#endregion
export { Sui as default };
