// @mysten/enoki/dist/EnokiFlow.mjs
import { EnokiClient } from "./EnokiClient/index.mjs";
import { createDefaultEncryption } from "./encryption.mjs";
import { EnokiKeypair } from "./EnokiKeypair.mjs";
import { createSessionStorage } from "./stores.mjs";
import { WebCryptoSigner } from "@mysten/signers/webcrypto";
import { decodeSuiPrivateKey } from "@mysten/sui/cryptography";
import { Ed25519Keypair } from "@mysten/sui/keypairs/ed25519";
import { fromBase64, toBase64 } from "@mysten/sui/utils";
import { decodeJwt } from "@mysten/sui/zklogin";
import { clear, createStore, get, set } from "idb-keyval";
import { atom, onMount, onSet } from "nanostores";

//#region src/EnokiFlow.ts
const createStorageKeys = (apiKey) => ({
	STATE: `@enoki/flow/state/${apiKey}`,
	SESSION: `@enoki/flow/session/${apiKey}`
});
/**
* @deprecated Use `registerEnokiWallets` instead
*/
var EnokiFlow = class {
	#storageKeys;
	#enokiClient;
	#encryption;
	#encryptionKey;
	#store;
	#useNativeCryptoSigner;
	#idbStore;
	constructor(config) {
		this.#enokiClient = new EnokiClient({
			apiKey: config.apiKey,
			apiUrl: config.apiUrl,
			additionalEpochs: config.additionalEpochs
		});
		this.#encryptionKey = config.apiKey;
		if (config.experimental_nativeCryptoSigner) {
			this.#useNativeCryptoSigner = true;
			this.#idbStore = createStore(config.apiKey, "enoki");
		} else this.#useNativeCryptoSigner = false;
		this.#encryption = config.encryption ?? createDefaultEncryption();
		this.#store = config.store ?? createSessionStorage();
		this.#storageKeys = createStorageKeys(config.apiKey);
		let storedState = null;
		try {
			const rawStoredValue = this.#store.get(this.#storageKeys.STATE);
			if (rawStoredValue) storedState = JSON.parse(rawStoredValue);
		} catch {}
		this.$zkLoginState = atom(storedState || {});
		this.$zkLoginSession = atom({
			initialized: false,
			value: null
		});
		onMount(this.$zkLoginSession, () => {
			this.getSession();
		});
		onSet(this.$zkLoginState, ({ newValue }) => {
			this.#store.set(this.#storageKeys.STATE, JSON.stringify(newValue));
		});
	}
	get enokiClient() {
		return this.#enokiClient;
	}
	async createAuthorizationURL(input) {
		const ephemeralKeyPair = this.#useNativeCryptoSigner ? await WebCryptoSigner.generate() : new Ed25519Keypair();
		const { nonce, randomness, maxEpoch, estimatedExpiration } = await this.#enokiClient.createZkLoginNonce({
			network: input.network,
			ephemeralPublicKey: ephemeralKeyPair.getPublicKey()
		});
		const params = new URLSearchParams({
			...input.extraParams,
			nonce,
			client_id: input.clientId,
			redirect_uri: input.redirectUrl,
			response_type: "id_token",
			scope: ["openid", ...input.extraParams && "scope" in input.extraParams ? input.extraParams.scope : []].filter(Boolean).join(" ")
		});
		let oauthUrl;
		switch (input.provider) {
			case "google":
				oauthUrl = `https://accounts.google.com/o/oauth2/v2/auth?${params}`;
				break;
			case "facebook":
				oauthUrl = `https://www.facebook.com/v17.0/dialog/oauth?${params}`;
				break;
			case "twitch":
				params.set("force_verify", "true");
				oauthUrl = `https://id.twitch.tv/oauth2/authorize?${params}`;
				break;
			default: throw new Error(`Invalid provider: ${input.provider}`);
		}
		this.$zkLoginState.set({ provider: input.provider });
		if (this.#useNativeCryptoSigner) await set("ephemeralKeyPair", ephemeralKeyPair.export(), this.#idbStore);
		await this.#setSession({
			expiresAt: estimatedExpiration,
			maxEpoch,
			randomness,
			ephemeralKeyPair: this.#useNativeCryptoSigner ? "@@native" : toBase64(decodeSuiPrivateKey(ephemeralKeyPair.getSecretKey()).secretKey)
		});
		return oauthUrl;
	}
	async handleAuthCallback(hash = window.location.hash) {
		const params = new URLSearchParams(hash.startsWith("#") ? hash.slice(1) : hash);
		const zkp = await this.getSession();
		if (!zkp || !zkp.ephemeralKeyPair || !zkp.maxEpoch || !zkp.randomness) throw new Error("Start of sign-in flow could not be found. Ensure you have started the sign-in flow before calling this.");
		const jwt = params.get("id_token");
		if (!jwt) throw new Error("Missing ID Token");
		decodeJwt(jwt);
		const { address, salt, publicKey } = await this.#enokiClient.getZkLogin({ jwt });
		this.$zkLoginState.set({
			...this.$zkLoginState.get(),
			salt,
			address,
			publicKey
		});
		await this.#setSession({
			...zkp,
			jwt
		});
		return params.get("state");
	}
	async #setSession(newValue) {
		if (newValue) {
			const storedValue = await this.#encryption.encrypt(this.#encryptionKey, JSON.stringify(newValue));
			this.#store.set(this.#storageKeys.SESSION, storedValue);
		} else this.#store.delete(this.#storageKeys.SESSION);
		this.$zkLoginSession.set({
			initialized: true,
			value: newValue
		});
	}
	async getSession() {
		if (this.$zkLoginSession.get().initialized) return this.$zkLoginSession.get().value;
		try {
			const storedValue = this.#store.get(this.#storageKeys.SESSION);
			if (!storedValue) return null;
			const state = JSON.parse(await this.#encryption.decrypt(this.#encryptionKey, storedValue));
			if (state?.expiresAt && Date.now() > state.expiresAt) await this.logout();
			else this.$zkLoginSession.set({
				initialized: true,
				value: state
			});
		} catch {
			this.$zkLoginSession.set({
				initialized: true,
				value: null
			});
		}
		return this.$zkLoginSession.get().value;
	}
	async logout() {
		this.$zkLoginState.set({});
		this.#store.delete(this.#storageKeys.STATE);
		if (this.#useNativeCryptoSigner) await clear(this.#idbStore);
		await this.#setSession(null);
	}
	async getProof({ network } = {}) {
		const zkp = await this.getSession();
		const { salt } = this.$zkLoginState.get();
		if (zkp?.proof) {
			if (zkp.expiresAt && Date.now() > zkp.expiresAt) throw new Error("Stored proof is expired.");
			return zkp.proof;
		}
		if (!salt || !zkp || !zkp.jwt) throw new Error("Missing required parameters for proof generation");
		let storedNativeSigner = void 0;
		if (this.#useNativeCryptoSigner && zkp.ephemeralKeyPair === "@@native") {
			storedNativeSigner = await get("ephemeralKeyPair", this.#idbStore);
			if (!storedNativeSigner) throw new Error("Native signer not found in store.");
		}
		const ephemeralKeyPair = zkp.ephemeralKeyPair === "@@native" ? WebCryptoSigner.import(storedNativeSigner) : Ed25519Keypair.fromSecretKey(fromBase64(zkp.ephemeralKeyPair));
		const proof = await this.#enokiClient.createZkLoginZkp({
			network,
			jwt: zkp.jwt,
			maxEpoch: zkp.maxEpoch,
			randomness: zkp.randomness,
			ephemeralPublicKey: ephemeralKeyPair.getPublicKey()
		});
		await this.#setSession({
			...zkp,
			proof
		});
		return proof;
	}
	async getKeypair({ network } = {}) {
		await this.getProof({ network });
		const zkp = await this.getSession();
		const { address } = this.$zkLoginState.get();
		if (!address || !zkp || !zkp.proof) throw new Error("Missing required data for keypair generation.");
		if (Date.now() > zkp.expiresAt) throw new Error("Stored proof is expired.");
		let storedNativeSigner = void 0;
		if (this.#useNativeCryptoSigner && zkp.ephemeralKeyPair === "@@native") {
			storedNativeSigner = await get("ephemeralKeyPair", this.#idbStore);
			if (!storedNativeSigner) throw new Error("Native signer not found in store.");
		}
		return new EnokiKeypair({
			address,
			ephemeralKeypair: zkp.ephemeralKeyPair === "@@native" ? WebCryptoSigner.import(storedNativeSigner) : Ed25519Keypair.fromSecretKey(fromBase64(zkp.ephemeralKeyPair)),
			maxEpoch: zkp.maxEpoch,
			proof: zkp.proof
		});
	}
};

//#endregion
export { EnokiFlow };
