// @mysten/enoki/dist/EnokiKeypair.mjs
import { Signer } from "@mysten/sui/cryptography";
import { ZkLoginPublicIdentifier, getZkLoginSignature } from "@mysten/sui/zklogin";

//#region src/EnokiKeypair.ts
var EnokiPublicKey = class extends ZkLoginPublicIdentifier {};
var EnokiKeypair = class extends Signer {
	#proof;
	#maxEpoch;
	#ephemeralKeypair;
	#publicKey;
	constructor(input) {
		super();
		this.#proof = input.proof;
		this.#maxEpoch = input.maxEpoch;
		this.#ephemeralKeypair = input.ephemeralKeypair;
		this.#publicKey = EnokiPublicKey.fromProof(input.address, input.proof);
	}
	async sign(data) {
		return this.#ephemeralKeypair.sign(data);
	}
	async signPersonalMessage(bytes) {
		const { bytes: signedBytes, signature: userSignature } = await this.#ephemeralKeypair.signPersonalMessage(bytes);
		return {
			bytes: signedBytes,
			signature: getZkLoginSignature({
				inputs: this.#proof,
				maxEpoch: this.#maxEpoch,
				userSignature
			})
		};
	}
	async signTransaction(bytes) {
		const { bytes: signedBytes, signature: userSignature } = await this.#ephemeralKeypair.signTransaction(bytes);
		return {
			bytes: signedBytes,
			signature: getZkLoginSignature({
				inputs: this.#proof,
				maxEpoch: this.#maxEpoch,
				userSignature
			})
		};
	}
	getKeyScheme() {
		return this.#ephemeralKeypair.getKeyScheme();
	}
	getPublicKey() {
		return this.#publicKey;
	}
};

//#endregion
export { EnokiKeypair, EnokiPublicKey };
