// @mysten/dapp-kit-react/dist/index.mjs
import { n as DAppKitContext, r as DAppKitProvider, t as useDAppKit } from "./useDAppKit-LUp-KcgG.mjs";
import { useStore } from "@nanostores/react";

export * from "@mysten/dapp-kit-core"

//#region src/hooks/useWallets.ts
function useWallets({ dAppKit } = {}) {
	return useStore(useDAppKit(dAppKit).stores.$wallets);
}

//#endregion
//#region src/hooks/useWalletConnection.ts
function useWalletConnection({ dAppKit } = {}) {
	return useStore(useDAppKit(dAppKit).stores.$connection);
}

//#endregion
//#region src/hooks/useCurrentAccount.ts
function useCurrentAccount({ dAppKit } = {}) {
	return useWalletConnection({ dAppKit }).account;
}

//#endregion
//#region src/hooks/useCurrentWallet.ts
function useCurrentWallet({ dAppKit } = {}) {
	return useWalletConnection({ dAppKit }).wallet;
}

//#endregion
//#region src/hooks/useCurrentClient.ts
function useCurrentClient({ dAppKit } = {}) {
	return useStore(useDAppKit(dAppKit).stores.$currentClient);
}

//#endregion
//#region src/hooks/useCurrentNetwork.ts
function useCurrentNetwork({ dAppKit } = {}) {
	return useStore(useDAppKit(dAppKit).stores.$currentNetwork);
}

//#endregion
export { DAppKitContext, DAppKitProvider, useCurrentAccount, useCurrentClient, useCurrentNetwork, useCurrentWallet, useDAppKit, useWalletConnection, useWallets };
