// @mysten/dapp-kit-react/dist/ui.mjs
import { t as useDAppKit } from "./useDAppKit-LUp-KcgG.mjs";
import * as React from "react";
import { jsx } from "react/jsx-runtime";
import { createComponent } from "@lit/react";
import { DAppKitConnectButton, DAppKitConnectModal } from "@mysten/dapp-kit-core/web";

//#region src/components/ConnectButton.tsx
const ConnectButtonComponent = createComponent({
	react: React,
	tagName: "mysten-dapp-kit-connect-button",
	elementClass: DAppKitConnectButton
});
function ConnectButton({ instance, ...props }) {
	const dAppKit = useDAppKit(instance);
	return /* @__PURE__ */ jsx(ConnectButtonComponent, {
		...props,
		instance: dAppKit
	});
}

//#endregion
//#region src/components/ConnectModal.tsx
const ConnectModalComponent = createComponent({
	react: React,
	tagName: "mysten-dapp-kit-connect-modal",
	elementClass: DAppKitConnectModal
});
function ConnectModal({ instance, ...props }) {
	const dAppKit = useDAppKit(instance);
	return /* @__PURE__ */ jsx(ConnectModalComponent, {
		...props,
		instance: dAppKit
	});
}

//#endregion
export { ConnectButton, ConnectModal };
