// @mysten/enoki/dist/EnokiClient/index.mjs
//#region src/EnokiClient/index.ts
const DEFAULT_API_URL = "https://api.enoki.mystenlabs.com";
const ZKLOGIN_HEADER = "zklogin-jwt";
var EnokiClientError = class extends Error {
	constructor(status, response) {
		let errors;
		try {
			errors = JSON.parse(response).errors;
		} catch {}
		const cause = errors?.[0] ? new Error(errors[0].message) : void 0;
		super(`Request to Enoki API failed (status: ${status})`, { cause });
		this.errors = [];
		this.errors = errors ?? [];
		this.name = "EnokiClientError";
		this.status = status;
		this.code = errors?.[0]?.code ?? "unknown_error";
	}
};
/**
* A low-level client for interacting with the Enoki API.
*/
var EnokiClient = class {
	#version;
	#apiUrl;
	#apiKey;
	#additionalEpochs;
	constructor(config) {
		this.#version = "v1";
		this.#apiUrl = config.apiUrl ?? DEFAULT_API_URL;
		this.#apiKey = config.apiKey;
		this.#additionalEpochs = config.additionalEpochs;
	}
	getApp(_input) {
		return this.#fetch("app", { method: "GET" });
	}
	getZkLogin(input) {
		return this.#fetch("zklogin", {
			method: "GET",
			headers: { [ZKLOGIN_HEADER]: input.jwt }
		});
	}
	getZkLoginAddresses(input) {
		return this.#fetch("zklogin/addresses", {
			method: "GET",
			headers: { [ZKLOGIN_HEADER]: input.jwt }
		});
	}
	createZkLoginNonce(input) {
		return this.#fetch("zklogin/nonce", {
			method: "POST",
			body: JSON.stringify({
				network: input.network,
				ephemeralPublicKey: input.ephemeralPublicKey.toSuiPublicKey(),
				additionalEpochs: input.additionalEpochs ?? this.#additionalEpochs
			})
		});
	}
	createZkLoginZkp(input) {
		return this.#fetch("zklogin/zkp", {
			method: "POST",
			headers: { [ZKLOGIN_HEADER]: input.jwt },
			body: JSON.stringify({
				network: input.network,
				ephemeralPublicKey: input.ephemeralPublicKey.toSuiPublicKey(),
				maxEpoch: input.maxEpoch,
				randomness: input.randomness
			})
		});
	}
	createSponsoredTransaction(input) {
		return this.#fetch("transaction-blocks/sponsor", {
			method: "POST",
			headers: input.jwt ? { [ZKLOGIN_HEADER]: input.jwt } : {},
			body: JSON.stringify({
				sender: input.sender,
				network: input.network,
				transactionBlockKindBytes: input.transactionKindBytes,
				allowedAddresses: input.allowedAddresses,
				allowedMoveCallTargets: input.allowedMoveCallTargets
			})
		});
	}
	executeSponsoredTransaction(input) {
		return this.#fetch(`transaction-blocks/sponsor/${input.digest}`, {
			method: "POST",
			body: JSON.stringify({ signature: input.signature })
		});
	}
	getSubnames(input) {
		const query = new URLSearchParams();
		if (input.address) query.set("address", input.address);
		if (input.network) query.set("network", input.network);
		if (input.domain) query.set("domain", input.domain);
		return this.#fetch("subnames" + (query.size > 0 ? `?${query.toString()}` : ""), { method: "GET" });
	}
	createSubname(input) {
		return this.#fetch("subnames", {
			method: "POST",
			headers: input.jwt ? { [ZKLOGIN_HEADER]: input.jwt } : {},
			body: JSON.stringify({
				network: input.network,
				domain: input.domain,
				subname: input.subname,
				targetAddress: input.targetAddress
			})
		});
	}
	deleteSubname(input) {
		this.#fetch("subnames", {
			method: "DELETE",
			body: JSON.stringify({
				network: input.network,
				domain: input.domain,
				subname: input.subname
			})
		});
	}
	async #fetch(path, init) {
		const res = await fetch(`${this.#apiUrl}/${this.#version}/${path}`, {
			...init,
			headers: {
				...init.headers,
				Authorization: `Bearer ${this.#apiKey}`,
				"Content-Type": "application/json",
				"Request-Id": crypto.randomUUID()
			}
		});
		if (!res.ok) throw new EnokiClientError(res.status, await res.text());
		const { data } = await res.json();
		return data;
	}
};

//#endregion
export { EnokiClient, EnokiClientError };
