// @mysten/signers/dist/aws/aws-kms-signer.mjs
import { getConcatenatedSignature } from "../utils/utils.mjs";
import { AwsKmsClient } from "./aws-client.mjs";
import { SIGNATURE_FLAG_TO_SCHEME, Signer } from "@mysten/sui/cryptography";
import { fromBase64, toBase64 } from "@mysten/sui/utils";

//#region src/aws/aws-kms-signer.ts
/**
* Aws KMS Signer integrates AWS Key Management Service (KMS) with the Sui blockchain
* to provide signing capabilities using AWS-managed cryptographic keys.
*/
var AwsKmsSigner = class AwsKmsSigner extends Signer {
	#publicKey;
	/** AWS KMS client instance */
	#client;
	/** AWS KMS Key ID used for signing */
	#kmsKeyId;
	/**
	* Creates an instance of AwsKmsSigner. It's expected to call the static `fromKeyId` method to create an instance.
	* For example:
	* ```
	* const signer = await AwsKmsSigner.fromKeyId(keyId, options);
	* ```
	* @throws Will throw an error if required AWS credentials or region are not provided.
	*/
	constructor({ kmsKeyId, client, publicKey }) {
		super();
		if (!kmsKeyId) throw new Error("KMS Key ID is required");
		this.#client = client;
		this.#kmsKeyId = kmsKeyId;
		this.#publicKey = publicKey;
	}
	/**
	* Retrieves the key scheme used by this signer.
	* @returns AWS supports only Secp256k1 and Secp256r1 schemes.
	*/
	getKeyScheme() {
		return SIGNATURE_FLAG_TO_SCHEME[this.#publicKey.flag()];
	}
	/**
	* Retrieves the public key associated with this signer.
	* @returns The Secp256k1PublicKey instance.
	* @throws Will throw an error if the public key has not been initialized.
	*/
	getPublicKey() {
		return this.#publicKey;
	}
	/**
	* Signs the given data using AWS KMS.
	* @param bytes - The data to be signed as a Uint8Array.
	* @returns A promise that resolves to the signature as a Uint8Array.
	* @throws Will throw an error if the public key is not initialized or if signing fails.
	*/
	async sign(bytes) {
		return getConcatenatedSignature(fromBase64((await this.#client.runCommand("Sign", {
			KeyId: this.#kmsKeyId,
			Message: toBase64(bytes),
			MessageType: "RAW",
			SigningAlgorithm: "ECDSA_SHA_256"
		})).Signature), this.getKeyScheme());
	}
	/**
	* Prepares the signer by fetching and setting the public key from AWS KMS.
	* It is recommended to initialize an `AwsKmsSigner` instance using this function.
	* @returns A promise that resolves once a `AwsKmsSigner` instance is prepared (public key is set).
	*/
	static async fromKeyId(keyId, options) {
		const client = new AwsKmsClient(options);
		return new AwsKmsSigner({
			kmsKeyId: keyId,
			client,
			publicKey: await client.getPublicKey(keyId)
		});
	}
};

//#endregion
export { AwsKmsSigner };
