// @ledgerhq/hw-bolos/lib/PKIError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PKITrustedNameTooSmallError = exports.PKIFailedVerificationError = exports.PKIMismatchKeyError = exports.PKIFailedHashError = exports.PKIUnknownCertificateTagError = exports.PKIUnknownDeviceError = exports.PKIIncorrectSignatureError = exports.PKIIncorrectCurveError = exports.PKIUnknownPublicUsageError = exports.PKIUnknownPublicIdError = exports.PKIUnknownSignerAlgorithmError = exports.PKIUnknownSignerIdError = exports.PKICertificateIndexError = exports.PKICertificateValidityError = exports.PKICertificateVersionError = exports.PKIStructError = exports.PKIError = exports.errorCodeValue = void 0;
exports.throwError = throwError;
const ErrorCode = {
    0x422f: "Incorrect structure type",
    0x4230: "Incorrect certificate version",
    0x4231: "Incorrect certificate validity",
    0x4232: "Incorrect certificate validity index",
    0x4233: "Unknown signer key ID",
    0x4234: "Unknown signature algorithm",
    0x4235: "Unknown public key ID",
    0x4236: "Unknown public key usage",
    0x4237: "Incorrect elliptic curve ID",
    0x4238: "Incorrect signature algorithm associated to the public key",
    0x4239: "Unknown target device",
    0x422d: "Unknown certificate tag",
    0x3301: "Failed to hash data",
    0x422e: "expected_key_usage doesn't match certificate key usage",
    0x5720: "Failed to verify signature",
    0x4118: "trusted_name buffer is too small to contain the trusted name",
};
exports.errorCodeValue = Object.keys(ErrorCode).map(v => parseInt(v));
function throwError(code) {
    switch (code) {
        case 0x422f:
            throw new PKIStructError();
        case 0x4230:
            throw new PKICertificateVersionError();
        case 0x4231:
            throw new PKICertificateValidityError();
        case 0x4232:
            throw new PKICertificateIndexError();
        case 0x4233:
            throw new PKIUnknownSignerIdError();
        case 0x4234:
            throw new PKIUnknownSignerAlgorithmError();
        case 0x4235:
            throw new PKIUnknownPublicIdError();
        case 0x4236:
            throw new PKIUnknownPublicUsageError();
        case 0x4237:
            throw new PKIIncorrectCurveError();
        case 0x4238:
            throw new PKIIncorrectSignatureError();
        case 0x4239:
            throw new PKIUnknownDeviceError();
        case 0x422d:
            throw new PKIUnknownCertificateTagError();
        case 0x3301:
            throw new PKIFailedHashError();
        case 0x422e:
            throw new PKIMismatchKeyError();
        case 0x5720:
            throw new PKIFailedVerificationError();
        case 0x4118:
            throw new PKITrustedNameTooSmallError();
        default:
            throw new PKIError("Unknown");
    }
}
class PKIError extends Error {
    constructor(cause) {
        super("PKIError due to " + cause);
    }
}
exports.PKIError = PKIError;
class PKIStructError extends PKIError {
    constructor() {
        super(ErrorCode[0x422f]);
    }
}
exports.PKIStructError = PKIStructError;
class PKICertificateVersionError extends PKIError {
    constructor() {
        super(ErrorCode[0x4230]);
    }
}
exports.PKICertificateVersionError = PKICertificateVersionError;
class PKICertificateValidityError extends PKIError {
    constructor() {
        super(ErrorCode[0x4231]);
    }
}
exports.PKICertificateValidityError = PKICertificateValidityError;
class PKICertificateIndexError extends PKIError {
    constructor() {
        super(ErrorCode[0x4232]);
    }
}
exports.PKICertificateIndexError = PKICertificateIndexError;
class PKIUnknownSignerIdError extends PKIError {
    constructor() {
        super(ErrorCode[0x4233]);
    }
}
exports.PKIUnknownSignerIdError = PKIUnknownSignerIdError;
class PKIUnknownSignerAlgorithmError extends PKIError {
    constructor() {
        super(ErrorCode[0x4234]);
    }
}
exports.PKIUnknownSignerAlgorithmError = PKIUnknownSignerAlgorithmError;
class PKIUnknownPublicIdError extends PKIError {
    constructor() {
        super(ErrorCode[0x4235]);
    }
}
exports.PKIUnknownPublicIdError = PKIUnknownPublicIdError;
class PKIUnknownPublicUsageError extends PKIError {
    constructor() {
        super(ErrorCode[0x4236]);
    }
}
exports.PKIUnknownPublicUsageError = PKIUnknownPublicUsageError;
class PKIIncorrectCurveError extends PKIError {
    constructor() {
        super(ErrorCode[0x4237]);
    }
}
exports.PKIIncorrectCurveError = PKIIncorrectCurveError;
class PKIIncorrectSignatureError extends PKIError {
    constructor() {
        super(ErrorCode[0x4238]);
    }
}
exports.PKIIncorrectSignatureError = PKIIncorrectSignatureError;
class PKIUnknownDeviceError extends PKIError {
    constructor() {
        super(ErrorCode[0x4239]);
    }
}
exports.PKIUnknownDeviceError = PKIUnknownDeviceError;
class PKIUnknownCertificateTagError extends PKIError {
    constructor() {
        super(ErrorCode[0x422d]);
    }
}
exports.PKIUnknownCertificateTagError = PKIUnknownCertificateTagError;
class PKIFailedHashError extends PKIError {
    constructor() {
        super(ErrorCode[0x3301]);
    }
}
exports.PKIFailedHashError = PKIFailedHashError;
class PKIMismatchKeyError extends PKIError {
    constructor() {
        super(ErrorCode[0x422e]);
    }
}
exports.PKIMismatchKeyError = PKIMismatchKeyError;
class PKIFailedVerificationError extends PKIError {
    constructor() {
        super(ErrorCode[0x5720]);
    }
}
exports.PKIFailedVerificationError = PKIFailedVerificationError;
class PKITrustedNameTooSmallError extends PKIError {
    constructor() {
        super(ErrorCode[0x4118]);
    }
}
exports.PKITrustedNameTooSmallError = PKITrustedNameTooSmallError;
