// @mysten/signers/dist/webcrypto/index.mjs
import { Signer } from "@mysten/sui/cryptography";
import { p256 } from "@noble/curves/nist.js";
import { Secp256r1PublicKey } from "@mysten/sui/keypairs/secp256r1";

//#region src/webcrypto/index.ts
function getCompressedPublicKey(publicKey) {
	const rawBytes = new Uint8Array(publicKey);
	const x = rawBytes.slice(1, 33);
	const prefix = (rawBytes.slice(33, 65)[31] & 1) === 0 ? 2 : 3;
	const compressed = new Uint8Array(Secp256r1PublicKey.SIZE);
	compressed[0] = prefix;
	compressed.set(x, 1);
	return compressed;
}
var WebCryptoSigner = class WebCryptoSigner extends Signer {
	#publicKey;
	static async generate({ extractable = false } = {}) {
		const keypair = await globalThis.crypto.subtle.generateKey({
			name: "ECDSA",
			namedCurve: "P-256"
		}, extractable, ["sign", "verify"]);
		const publicKey = await globalThis.crypto.subtle.exportKey("raw", keypair.publicKey);
		return new WebCryptoSigner(keypair.privateKey, getCompressedPublicKey(new Uint8Array(publicKey)));
	}
	/**
	* Imports a keypair using the value returned by `export()`.
	*/
	static import(data) {
		return new WebCryptoSigner(data.privateKey, data.publicKey);
	}
	getKeyScheme() {
		return "Secp256r1";
	}
	constructor(privateKey, publicKey) {
		super();
		this.privateKey = privateKey;
		this.#publicKey = new Secp256r1PublicKey(publicKey);
	}
	/**
	* Exports the keypair so that it can be stored in IndexedDB.
	*/
	export() {
		const exportedKeypair = {
			privateKey: this.privateKey,
			publicKey: this.#publicKey.toRawBytes()
		};
		Object.defineProperty(exportedKeypair, "toJSON", {
			enumerable: false,
			value: () => {
				throw new Error("The exported keypair must not be serialized. It must be stored in IndexedDB directly.");
			}
		});
		return exportedKeypair;
	}
	getPublicKey() {
		return this.#publicKey;
	}
	async sign(bytes) {
		const rawSignature = await globalThis.crypto.subtle.sign({
			name: "ECDSA",
			hash: "SHA-256"
		}, this.privateKey, bytes);
		const signature = p256.Signature.fromBytes(new Uint8Array(rawSignature));
		return (signature.hasHighS() ? new p256.Signature(signature.r, p256.Point.Fn.neg(signature.s)) : signature).toBytes("compact");
	}
};

//#endregion
export { WebCryptoSigner };
