// @ledgerhq/live-network/lib/cache.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeLRUCache = void 0;
exports.seconds = seconds;
exports.minutes = minutes;
exports.hours = hours;
const lru_cache_1 = __importDefault(require("lru-cache"));
function seconds(num, max = 100) {
    return {
        max,
        ttl: num * 1000,
    };
}
function minutes(num, max = 100) {
    return seconds(num * 60, max);
}
function hours(num, max = 100) {
    return minutes(num * 60, max);
}
/**
 * Creates an LRU cache function that wraps the given async function `f`.
 * By default it will have a lifespan of 300 seconds and a change in the key
 * will invalidate its value.
 *
 * @param {(...args: A) => Promise<T>} f - The async function to be cached.
 * @param {(...args: A) => string} [keyExtractor] - A function that extracts a cache key from the arguments passed to `f`.
 * @param {LRU.Options<string, any>} [lruOpts] - Options for the internal LRU cache.
 *
 * @returns {CacheRes<A, T>} A cached version of the async function `f`.
 */
const makeLRUCache = (f, keyExtractor = () => "", lruOpts = {
    max: 100,
    ttl: 5 * 60 * 1000,
}) => {
    // LRU-Cache is written in JS and do not enforce in its code the type checking.
    // Regarding its [documentation](https://github.com/isaacs/node-lru-cache/#ttl), `max` or `ttlAutopurge` must be set.
    // As the code in live use sometimes `max` property with `ttl`, we check it's defined in `lruOpts` to add or not `ttlAutopurge`.
    // eslint-disable-next-line
    // @ts-ignore: TS-2339
    lruOpts = lruOpts.max
        ? lruOpts
        : {
            ...lruOpts,
            ttlAutopurge: true,
        };
    const cache = new lru_cache_1.default(lruOpts);
    const result = (...args) => {
        const key = keyExtractor(...args);
        let promise = cache.get(key);
        if (promise)
            return promise;
        promise = f(...args).catch(e => {
            cache.delete(key);
            throw e;
        });
        cache.set(key, promise);
        return promise;
    };
    result.force = (...args) => {
        const key = keyExtractor(...args);
        const promise = f(...args).catch(e => {
            cache.delete(key);
            throw e;
        });
        cache.set(key, promise);
        return promise;
    };
    result.hydrate = (key, value) => {
        cache.set(key, Promise.resolve(value));
    };
    result.clear = (key) => {
        cache.delete(key);
    };
    result.reset = () => {
        cache.clear();
    };
    return result;
};
exports.makeLRUCache = makeLRUCache;
