// @mysten/enoki/dist/wallet/state.mjs
import { createDefaultEncryption } from "../encryption.mjs";
import { clear, createStore, del, get, set } from "idb-keyval";
import { atom, onMount, onSet, task } from "nanostores";

//#region src/wallet/state.ts
const sessionKey = "zklogin-session";
const stateKey = "zklogin-state";
var EnokiWalletState = class {
	#encryption;
	#encryptionKey;
	#stateStore;
	#sessionContextByNetwork;
	#zkLoginState;
	#hydrated = null;
	constructor(config) {
		this.#encryptionKey = config.apiKey;
		this.#encryption = createDefaultEncryption();
		this.#stateStore = createStore(`${config.apiKey}_${config.clientId}`, "enoki");
		this.#zkLoginState = atom(null);
		this.#initZkLoginState();
		this.#sessionContextByNetwork = config.clients.reduce((accumulator, client) => {
			const network = client.network;
			const idbStore = createStore(`${config.apiKey}_${network}_${config.clientId}`, "enoki");
			const sessionContext = {
				$zkLoginSession: atom({
					initialized: false,
					value: null
				}),
				client,
				idbStore
			};
			onMount(sessionContext.$zkLoginSession, () => {
				this.getSession(sessionContext);
			});
			return accumulator.set(network, sessionContext);
		}, /* @__PURE__ */ new Map());
	}
	get zkLoginState() {
		return this.#zkLoginState;
	}
	ensureHydrated() {
		return this.#hydrateFromStorage();
	}
	get sessionContextByNetwork() {
		return this.#sessionContextByNetwork;
	}
	getSessionContext(network) {
		const context = this.#sessionContextByNetwork.get(network);
		if (!context) throw new Error(`The network ${network} isn't supported.`);
		return context;
	}
	async logout() {
		this.#zkLoginState.set(null);
		await clear(this.#stateStore);
		for (const context of this.#sessionContextByNetwork.values()) {
			await this.setSession(context, null);
			await clear(context.idbStore);
		}
	}
	async setSession(context, newValue) {
		if (newValue) await set(sessionKey, await this.#encryption.encrypt(this.#encryptionKey, JSON.stringify(newValue)), context.idbStore);
		else await del(sessionKey, context.idbStore);
		context.$zkLoginSession.set({
			initialized: true,
			value: newValue
		});
	}
	async getSession({ $zkLoginSession, idbStore }) {
		if ($zkLoginSession.get().initialized) return $zkLoginSession.get().value;
		try {
			const storedValue = await get(sessionKey, idbStore);
			if (!storedValue) return null;
			const state = JSON.parse(await this.#encryption.decrypt(this.#encryptionKey, storedValue));
			if (state?.expiresAt && Date.now() > state.expiresAt) await this.logout();
			else $zkLoginSession.set({
				initialized: true,
				value: state
			});
		} catch {
			$zkLoginSession.set({
				initialized: true,
				value: null
			});
		}
		return $zkLoginSession.get().value;
	}
	#initZkLoginState() {
		onSet(this.#zkLoginState, ({ newValue }) => {
			set(stateKey, JSON.stringify(newValue), this.#stateStore);
		});
		onMount(this.#zkLoginState, () => {
			this.#hydrateFromStorage();
		});
	}
	#hydrateFromStorage() {
		if (!this.#hydrated) this.#hydrated = task(async () => {
			try {
				const rawStoredValue = await get(stateKey, this.#stateStore);
				if (rawStoredValue) this.#zkLoginState.set(JSON.parse(rawStoredValue));
			} catch {}
		});
		return this.#hydrated;
	}
};

//#endregion
export { EnokiWalletState };
