// @mysten/enoki/dist/wallet/utils.mjs
import { EnokiGetMetadata, EnokiGetSession } from "./features.mjs";
import { getWalletFeature } from "@wallet-standard/ui";

//#region src/wallet/utils.ts
function isEnokiWallet(wallet) {
	if (isWalletHandle(wallet)) return wallet.features.includes(EnokiGetMetadata);
	return EnokiGetMetadata in wallet.features;
}
function getWalletMetadata(wallet) {
	if (isWalletHandle(wallet)) try {
		const { getMetadata } = getWalletFeature(wallet, EnokiGetMetadata);
		return getMetadata();
	} catch {
		return null;
	}
	else if (EnokiGetMetadata in wallet.features) return wallet.features[EnokiGetMetadata].getMetadata();
	return null;
}
async function getSession(wallet, input) {
	if (isWalletHandle(wallet)) try {
		const { getSession: getSession$1 } = getWalletFeature(wallet, EnokiGetSession);
		return await getSession$1(input);
	} catch {
		return null;
	}
	else if (EnokiGetSession in wallet.features) return await wallet.features[EnokiGetSession].getSession(input);
	return null;
}
function isGoogleWallet(wallet) {
	return getWalletMetadata(wallet)?.provider === "google";
}
function isTwitchWallet(wallet) {
	return getWalletMetadata(wallet)?.provider === "twitch";
}
function isFacebookWallet(wallet) {
	return getWalletMetadata(wallet)?.provider === "facebook";
}
function isWalletHandle(wallet) {
	return Array.isArray(wallet.features);
}

//#endregion
export { getSession, getWalletMetadata, isEnokiWallet, isFacebookWallet, isGoogleWallet, isTwitchWallet };
