// @mysten/signers/dist/ledger/index.mjs
import { getInputObjects } from "./objects.mjs";
import { Signer, messageWithIntent, toSerializedSignature } from "@mysten/sui/cryptography";
import { toBase64 } from "@mysten/sui/utils";
import { Ed25519PublicKey } from "@mysten/sui/keypairs/ed25519";
import { Transaction } from "@mysten/sui/transactions";
import { bcs } from "@mysten/sui/bcs";

//#region src/ledger/index.ts
/**
* Ledger integrates with the Sui blockchain to provide signing capabilities using Ledger devices.
*/
var LedgerSigner = class LedgerSigner extends Signer {
	#derivationPath;
	#publicKey;
	#ledgerClient;
	#suiClient;
	/**
	* Creates an instance of LedgerSigner. It's expected to call the static `fromDerivationPath` method to create an instance.
	* @example
	* ```
	* const signer = await LedgerSigner.fromDerivationPath(derivationPath, options);
	* ```
	*/
	constructor({ publicKey, derivationPath, ledgerClient, suiClient }) {
		super();
		this.#publicKey = publicKey;
		this.#derivationPath = derivationPath;
		this.#ledgerClient = ledgerClient;
		this.#suiClient = suiClient;
	}
	/**
	* Retrieves the key scheme used by this signer.
	*/
	getKeyScheme() {
		return "ED25519";
	}
	/**
	* Retrieves the public key associated with this signer.
	* @returns The Ed25519PublicKey instance.
	*/
	getPublicKey() {
		return this.#publicKey;
	}
	/**
	* Signs the provided transaction bytes.
	* @returns The signed transaction bytes and signature.
	*/
	async signTransaction(bytes, bcsObjects, resolution) {
		const transactionOptions = bcsObjects ? { bcsObjects } : await getInputObjects(Transaction.from(bytes), this.#suiClient).catch(() => ({ bcsObjects: [] }));
		const intentMessage = messageWithIntent("TransactionData", bytes);
		const { signature } = await this.#ledgerClient.signTransaction(this.#derivationPath, intentMessage, transactionOptions, resolution);
		return {
			bytes: toBase64(bytes),
			signature: toSerializedSignature({
				signature,
				signatureScheme: this.getKeyScheme(),
				publicKey: this.#publicKey
			})
		};
	}
	/**
	* Signs the provided personal message.
	* @returns The signed message bytes and signature.
	*/
	async signPersonalMessage(bytes) {
		const intentMessage = messageWithIntent("PersonalMessage", bcs.byteVector().serialize(bytes).toBytes());
		const { signature } = await this.#ledgerClient.signTransaction(this.#derivationPath, intentMessage);
		return {
			bytes: toBase64(bytes),
			signature: toSerializedSignature({
				signature,
				signatureScheme: this.getKeyScheme(),
				publicKey: this.#publicKey
			})
		};
	}
	/**
	* Prepares the signer by fetching and setting the public key from a Ledger device.
	* It is recommended to initialize an `LedgerSigner` instance using this function.
	* @returns A promise that resolves once a `LedgerSigner` instance is prepared (public key is set).
	*/
	static async fromDerivationPath(derivationPath, ledgerClient, suiClient) {
		const { publicKey } = await ledgerClient.getPublicKey(derivationPath);
		if (!publicKey) throw new Error("Failed to get public key from Ledger.");
		return new LedgerSigner({
			derivationPath,
			publicKey: new Ed25519PublicKey(publicKey),
			ledgerClient,
			suiClient
		});
	}
	/**
	* Generic signing is not supported by Ledger.
	* @throws Always throws an error indicating generic signing is unsupported.
	*/
	sign() {
		throw new Error("Ledger Signer does not support generic signing.");
	}
	/**
	* Generic signing is not supported by Ledger.
	* @throws Always throws an error indicating generic signing is unsupported.
	*/
	signWithIntent() {
		throw new Error("Ledger Signer does not support generic signing.");
	}
};

//#endregion
export { LedgerSigner, getInputObjects };
