// @ledgerhq/ledger-cal-service/lib/certificate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertCertificateToDeviceData = convertCertificateToDeviceData;
exports.getCertificate = getCertificate;
const live_network_1 = __importDefault(require("@ledgerhq/live-network"));
const common_1 = require("./common");
const live_env_1 = require("@ledgerhq/live-env");
const DeviceModel = {
    blue: "blue",
    nanoS: "nanos",
    nanoSP: "nanosp",
    nanoX: "nanox",
    stax: "stax",
    /** Ledger Flex ("europa" is the internal name) */
    europa: "flex",
    flex: "flex",
    apex: "apexp",
};
function hexToUint8Array(hex) {
    if (hex.length % 2 !== 0) {
        throw new Error(`Invalid hex string length: expected even length, got ${hex.length}`);
    }
    if (!/^[0-9a-fA-F]*$/.test(hex)) {
        throw new Error("Invalid hex string content: only [0-9a-fA-F] characters are allowed");
    }
    const bytes = new Uint8Array(hex.length / 2);
    for (let i = 0; i < bytes.length; i++) {
        bytes[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }
    return bytes;
}
const SIGNATURE_TAG = 0x15;
const MAX_SIGNATURE_LENGTH = 0xff;
function convertCertificateToDeviceData(info) {
    const descriptorBytes = hexToUint8Array(info.descriptor);
    const signatureBytes = hexToUint8Array(info.signature);
    if (signatureBytes.length > MAX_SIGNATURE_LENGTH) {
        throw new Error(`Signature too long: ${signatureBytes.length} bytes (maximum ${MAX_SIGNATURE_LENGTH})`);
    }
    const result = new Uint8Array(descriptorBytes.length + 2 + signatureBytes.length);
    let offset = 0;
    result.set(descriptorBytes, offset);
    offset += descriptorBytes.length;
    result[offset++] = SIGNATURE_TAG;
    result[offset++] = signatureBytes.length;
    result.set(signatureBytes, offset);
    return result;
}
function publicKeyIdOf(usage) {
    switch (usage) {
        case "perps_data":
            return "yield";
        case "trusted_name":
            return "domain_metadata_key";
        case "coin_meta":
            return "token_metadata_key";
    }
}
/**
 * Retrieve PKI certificate
 * @param device
 */
async function getCertificate(device, usage, version = "latest", { env = "prod", signatureKind = "prod", ref = (0, live_env_1.getEnv)("CAL_REF") || undefined, } = common_1.DEFAULT_OPTION) {
    let params = {
        output: "id,target_device,not_valid_after,public_key_usage,certificate_version,descriptor",
        target_device: DeviceModel[device],
        public_key_usage: usage,
        public_key_id: publicKeyIdOf(usage),
        ref,
    };
    if (version === "latest") {
        params = {
            ...params,
            latest: true,
        };
    }
    else {
        params = {
            ...params,
            not_valid_after: version,
        };
    }
    const { data } = await (0, live_network_1.default)({
        method: "GET",
        url: `${(0, common_1.getCALDomain)(env)}/v1/certificates`,
        params,
    });
    if (data.length === 0) {
        throw new Error("Empty result");
    }
    return {
        descriptor: data[0].descriptor.data,
        signature: data[0].descriptor.signatures[signatureKind],
    };
}
