// @ledgerhq/ledger-cal-service/lib/index.js
"use strict";
/*
 * Crypto Asset List Service
 *
 * Use only exposed methods below outside of this module.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SWAP_DATA_CDN = exports.convertCertificateToDeviceData = void 0;
const certificate_1 = require("./certificate");
const currencies_1 = require("./currencies");
const networks_1 = require("./networks");
const partners_1 = require("./partners");
const tokens_1 = require("./tokens");
var certificate_2 = require("./certificate");
Object.defineProperty(exports, "convertCertificateToDeviceData", { enumerable: true, get: function () { return certificate_2.convertCertificateToDeviceData; } });
var default_1 = require("./partners/default");
Object.defineProperty(exports, "SWAP_DATA_CDN", { enumerable: true, get: function () { return default_1.SWAP_DATA_CDN; } });
exports.default = {
    findCurrencyData: currencies_1.findCurrencyData,
    findToken: tokens_1.findCachedToken,
    getCertificate: certificate_1.getCertificate,
    getNetworks: networks_1.getCachedNetworks,
    getProvidersCDNData: partners_1.getProvidersCDNData,
    getProvidersData: partners_1.getProvidersData,
};
